web/src/narrow.js

```diff
@@ -94,6 +94,56 @@ export function save_narrow(terms, trigger) {
     changehash(new_hash, trigger);
 }
 
+function change_view_for_narrow_containing_with_operator(message_data, opts) {
+    let anchor;
+    const msg_ids = message_data.messages.map((msg) => msg.id);
+    const msgs = msg_ids.map((msg_id) => message_store.get(msg_id));
+
+    if (message_data.anchor === LARGER_THAN_MAX_MESSAGE_ID) {
+        anchor = Math.max(...msg_ids);
+    } else {
+        anchor = message_data.anchor;
+    }
+    const message = msgs[0];
+    assert(message.type === "stream");
+    const channel_name = message.display_recipient;
+    const topic_name = message.topic;
+
+    const adjusted_terms = [
+        {operator: "stream", operand: channel_name, negated: false},
+        {operator: "topic", operand: topic_name, negated: false},
+        {operator: "with", operand: `${anchor}`, negated: false},
+    ];
+
+    const filter = new Filter(adjusted_terms);
+    const excludes_muted_topics = filter.excludes_muted_topics();
+    const msg_data = new MessageListData({
+        filter,
+        excludes_muted_topics,
+    });
+
+    const id_info = {
+        target_id: anchor,
+        local_select_id: anchor,
+        final_select_id: anchor,
+    };
+
+    maybe_add_local_messages({
+        id_info,
+        msg_data,
+    });
+
+    msg_data.add_messages(msgs);
+    save_narrow(adjusted_terms, opts.trigger);
+
+    const msg_list = new message_list.MessageList({
+        data: msg_data,
+    });
+
+    message_lists.update_current_message_list(msg_list);
+    handle_post_view_change(msg_list);
+}
```

Comment:
Is there a way, perhaps involving some extracted functions, that this can share code with the main narrowing code path? I worry a lot that this duplicates a bunch of fiddly logic and will bitrot into incorrectness quickly.

Reply:
Prep commit - #30172

Reply:
updated to use the common method.